File: frontend/src/api/canvas.ts
import { apiDelete, apiGet, apiPatch, apiPost } from './client';

// ---------------------------------------------------------------------------
// Backend schema types (Phase 7 — Visual Thinking Canvas)
// ---------------------------------------------------------------------------

export type CanvasItemType = 'note' | 'group' | 'text';
export type CanvasEdgeStyle = 'default' | 'dashed' | 'bold';

export interface CanvasOut {
  id: string;
  title: string;
  description: string | null;
  viewport_x: number;
  viewport_y: number;
  viewport_zoom: number;
  item_count: number;
  created_at: string;
  updated_at: string;
}

export interface CanvasItemOut {
  id: string;
  canvas_id: string;
  note_id: string | null;
  item_type: CanvasItemType;
  position_x: number;
  position_y: number;
  width: number | null;
  height: number | null;
  color: string | null;
  label: string | null;
  z_index: number;
  version: number;
  last_known_title: string | null;
  note_title: string | null;
  note_summary: string | null;
  note_content: string | null;
  created_at: string;
  updated_at: string;
}

export interface CanvasEdgeOut {
  id: string;
  canvas_id: string;
  source_item_id: string;
  target_item_id: string;
  label: string | null;
  style: CanvasEdgeStyle;
  created_at: string;
}

export interface CanvasDetailOut extends CanvasOut {
  items: CanvasItemOut[];
  edges: CanvasEdgeOut[];
}

export interface CanvasListResponse {
  items: CanvasOut[];
  total: number;
}

export interface CanvasCreate {
  title?: string;
  description?: string;
}

export interface CanvasUpdate {
  title?: string;
  description?: string;
  viewport_x?: number;
  viewport_y?: number;
  viewport_zoom?: number;
}

export interface CanvasItemCreate {
  note_id?: string;
  item_type: CanvasItemType;
  position_x?: number;
  position_y?: number;
  width?: number;
  height?: number;
  color?: string;
  label?: string;
  z_index?: number;
}

export interface CanvasItemUpdate {
  position_x?: number;
  position_y?: number;
  width?: number;
  height?: number;
  color?: string;
  label?: string;
  z_index?: number;
  version: number;
}

export interface BatchItemEntry {
  id: string;
  position_x?: number;
  position_y?: number;
  width?: number;
  height?: number;
  z_index?: number;
  version: number;
}

export interface CanvasEdgeCreate {
  source_item_id: string;
  target_item_id: string;
  label?: string;
  style?: CanvasEdgeStyle;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function listCanvases(): Promise<CanvasListResponse> {
  return apiGet<CanvasListResponse>('/api/canvases');
}

export async function createCanvas(data: CanvasCreate = {}): Promise<CanvasOut> {
  return apiPost<CanvasOut>('/api/canvases', data);
}

export async function getCanvas(id: string): Promise<CanvasDetailOut> {
  return apiGet<CanvasDetailOut>(`/api/canvases/${id}`);
}

export async function updateCanvas(id: string, patch: CanvasUpdate): Promise<CanvasOut> {
  return apiPatch<CanvasOut>(`/api/canvases/${id}`, patch);
}

export async function deleteCanvas(id: string): Promise<void> {
  return apiDelete(`/api/canvases/${id}`);
}

export async function addCanvasItem(
  canvasId: string,
  item: CanvasItemCreate,
): Promise<CanvasItemOut> {
  return apiPost<CanvasItemOut>(`/api/canvases/${canvasId}/items`, item);
}

export async function updateCanvasItem(
  canvasId: string,
  itemId: string,
  patch: CanvasItemUpdate,
): Promise<CanvasItemOut> {
  return apiPatch<CanvasItemOut>(`/api/canvases/${canvasId}/items/${itemId}`, patch);
}

export async function batchUpdateItems(
  canvasId: string,
  items: BatchItemEntry[],
): Promise<CanvasItemOut[]> {
  const resp = await apiPost<{ items: CanvasItemOut[] }>(`/api/canvases/${canvasId}/items/batch`, { items });
  return resp.items;
}

export async function deleteCanvasItem(canvasId: string, itemId: string): Promise<void> {
  return apiDelete(`/api/canvases/${canvasId}/items/${itemId}`);
}

export async function addCanvasEdge(
  canvasId: string,
  edge: CanvasEdgeCreate,
): Promise<CanvasEdgeOut> {
  return apiPost<CanvasEdgeOut>(`/api/canvases/${canvasId}/edges`, edge);
}

export async function deleteCanvasEdge(canvasId: string, edgeId: string): Promise<void> {
  return apiDelete(`/api/canvases/${canvasId}/edges/${edgeId}`);
}

export async function autoLayoutCanvas(canvasId: string): Promise<CanvasItemOut[]> {
  const resp = await apiPost<{ items: CanvasItemOut[] }>(`/api/canvases/${canvasId}/auto-layout`, {});
  return resp.items;
}

File: backend/app/schemas/canvas.py
"""
Pydantic schemas for Phase 7 Visual Thinking Canvas endpoints.
"""
import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


_ITEM_TYPE = Literal["note", "group", "text"]
_EDGE_STYLE = Literal["default", "dashed", "bold"]


# ---------------------------------------------------------------------------
# Canvas — request
# ---------------------------------------------------------------------------

class CanvasCreate(BaseModel):
    title: str = Field(default="Untitled Canvas", max_length=200)
    description: Optional[str] = None


class CanvasUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=200)
    description: Optional[str] = None
    viewport_x: Optional[float] = None
    viewport_y: Optional[float] = None
    viewport_zoom: Optional[float] = Field(default=None, ge=0.1, le=5.0)


# ---------------------------------------------------------------------------
# Canvas items — request
# ---------------------------------------------------------------------------

class CanvasItemCreate(BaseModel):
    note_id: Optional[uuid.UUID] = None
    item_type: _ITEM_TYPE
    position_x: float = 0.0
    position_y: float = 0.0
    width: Optional[float] = Field(default=None, gt=0)
    height: Optional[float] = Field(default=None, gt=0)
    color: Optional[str] = Field(default=None, max_length=20)
    label: Optional[str] = None
    z_index: int = 0


class CanvasItemUpdate(BaseModel):
    """Single-item PATCH. `version` is REQUIRED for optimistic concurrency."""
    position_x: Optional[float] = None
    position_y: Optional[float] = None
    width: Optional[float] = Field(default=None, gt=0)
    height: Optional[float] = Field(default=None, gt=0)
    color: Optional[str] = Field(default=None, max_length=20)
    label: Optional[str] = None
    z_index: Optional[int] = None
    version: int


class BatchItemEntry(BaseModel):
    """One entry inside a batch position-update request."""
    id: uuid.UUID
    position_x: Optional[float] = None
    position_y: Optional[float] = None
    width: Optional[float] = Field(default=None, gt=0)
    height: Optional[float] = Field(default=None, gt=0)
    z_index: Optional[int] = None
    version: int


class BatchItemUpdateRequest(BaseModel):
    items: list[BatchItemEntry]


# ---------------------------------------------------------------------------
# Edges — request
# ---------------------------------------------------------------------------

class CanvasEdgeCreate(BaseModel):
    source_item_id: uuid.UUID
    target_item_id: uuid.UUID
    label: Optional[str] = None
    style: _EDGE_STYLE = "default"


# ---------------------------------------------------------------------------
# Response
# ---------------------------------------------------------------------------

class CanvasItemOut(BaseModel):
    id: uuid.UUID
    canvas_id: uuid.UUID
    note_id: Optional[uuid.UUID] = None
    item_type: str
    position_x: float
    position_y: float
    width: Optional[float] = None
    height: Optional[float] = None
    color: Optional[str] = None
    label: Optional[str] = None
    z_index: int
    version: int
    last_known_title: Optional[str] = None
    # Inlined for the frontend's render-from-canvas-payload path.
    note_title: Optional[str] = None
    note_summary: Optional[str] = None
    note_content: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CanvasEdgeOut(BaseModel):
    id: uuid.UUID
    canvas_id: uuid.UUID
    source_item_id: uuid.UUID
    target_item_id: uuid.UUID
    label: Optional[str] = None
    style: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CanvasOut(BaseModel):
    id: uuid.UUID
    title: str
    description: Optional[str] = None
    viewport_x: float
    viewport_y: float
    viewport_zoom: float
    item_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CanvasDetailOut(CanvasOut):
    items: list[CanvasItemOut] = []
    edges: list[CanvasEdgeOut] = []


class CanvasListResponse(BaseModel):
    items: list[CanvasOut]
    total: int


class BatchUpdateResponse(BaseModel):
    items: list[CanvasItemOut]
